nav2_dwb_controller/dwb_critics/test/map_grid_test.cpp

```diff
@@ -0,0 +1,238 @@
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_global_costmap");
+  auto node = nav2_util::LifecycleNode::make_shared("costmap_tester");
+  node->configure();
+  node->activate();
+  costmap_ros->configure();
+
+  std::string name = "test";
+  std::string ns = "ns";
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, ns + "." + name + ".aggregation_type",
+    rclcpp::ParameterValue(std::string("product")));
+
+  critic->initialize(node, name, ns, costmap_ros);
+
+  EXPECT_TRUE(critic->aggregationTypeIsProduct());
+}
+
+TEST(MapGrid, InvalidValue)
+{
+  auto critic = std::make_shared<OpenMapGrid>();
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_global_costmap");
+  auto node = nav2_util::LifecycleNode::make_shared("costmap_tester");
+  node->configure();
+  node->activate();
+  costmap_ros->configure();
+
+  std::string name = "test";
+  std::string ns = "ns";
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, ns + "." + name + ".aggregation_type",
+    rclcpp::ParameterValue(std::string("invalid_value")));
+
+  critic->initialize(node, name, ns, costmap_ros);
+
+  // For an invalid type it should fall back to Last
+  EXPECT_TRUE(critic->aggregationTypeIsLast());
+}
+
+TEST(MapGrid, ObstacleUnreachable)
+{
+  auto critic = std::make_shared<OpenMapGrid>();
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_global_costmap");
+  auto node = nav2_util::LifecycleNode::make_shared("costmap_tester");
+  node->configure();
+  node->activate();
+  costmap_ros->configure();
+
+  std::string name = "test";
+  std::string ns = "ns";
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, ns + "." + name + ".aggregation_type",
+    rclcpp::ParameterValue(std::string("last")));
+
+  critic->initialize(node, name, ns, costmap_ros);
+  critic->reset();
+
+  // The value for obstacle should be the number of grid cell
+  unsigned int size_x = costmap_ros->getCostmap()->getSizeInCellsX();
+  unsigned int size_y = costmap_ros->getCostmap()->getSizeInCellsY();
+  unsigned int obstacle_cost = size_x * size_y;
+  unsigned int unreachable_cost = obstacle_cost + 1;
+
+  std::cout << "Size (x * y) " << unreachable_cost << std::endl;
+  critic->setAsObstacle(2);
+  EXPECT_EQ(critic->getValue(1), unreachable_cost);
+  EXPECT_EQ(critic->getValue(2), obstacle_cost);
+  EXPECT_EQ(critic->getValue(3), unreachable_cost);
+}
+
+TEST(MapGrid, CriticVisualization)
+{
+  auto critic = std::make_shared<OpenMapGrid>();
+
+  auto node = nav2_util::LifecycleNode::make_shared("base_obstacle_critic_tester");
+  node->configure();
+  node->activate();
+
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_global_costmap");
+  costmap_ros->configure();
+
+  std::string name = "name";
+  std::string ns = "ns";
+
+  critic->initialize(node, name, ns, costmap_ros);
+  // This makes all cells "unreachable"
+  critic->reset();
```

Comment:
It might be good to change the value of 1-2 things so that as you iterate you can validate that it reset everything but respected setting 1-2 things. 

Or set 1-2 things, then reset, and check that they properly cleared.

Reply:
I tried these and I could not manage to change some values (even by changing them in the costmap directly).

Reply:
Didn't you do so in the previous test using setAsObstacle`? I think the prop function also will update a full field of these values - one of the functions you didn't call in these tests

Reply:
Yes, indeed. setAsObstacle is the only thing that changes values from unreachable to obstacle. I will use that one, to make some changes in the map.